### PR TITLE
match gemm_softmax_gemm when there is no scale

### PR DESCRIPTION
**Pull request by shivadbhavsar**
Make the scale mul op optional when looking for attention. Required to get mlir attention for for SDXL workflow.

### CODE REVIEW

File: src/targets/gpu/prefuse_ops.cpp [hunk 1]
```diff
@@ -181,17 +182,20 @@ struct find_gemm_softmax_gemm
...
         float scale = 1.0;
-        scale_lit->eval().visit([&](const auto s) {
-            // CK only supports single-valued scale
-            if(std::all_of(
-                   s.begin() + 1, s.end(), [&](auto v) { return float_equal(v, s.front()); }))
-                scale = s.front();
-            else
-                return;
-        });
+        if(r.instructions.find("scale") != r.instructions.end())
```
Comment (pfultz2) [inline src/targets/gpu/prefuse_ops.cpp:187]: Use `contains(r.instructions, "scale")` instead.

File: src/targets/gpu/prefuse_ops.cpp [hunk 2]
```diff
@@ -181,17 +182,20 @@ struct find_gemm_softmax_gemm
...
-                scale = s.front();
-            else
-                return;
-        });
+        if(contains(r.instructions, "scale"))
+        {
+            auto scale_lit = r.instructions["scale"];
+            scale_lit->eval().visit([&](const auto s) {
+                // CK only supports single-valued scale
+                if(std::all_of(
```
Comment (kahmed10) [inline src/targets/gpu/prefuse_ops.cpp:192]: It might be better to flip the logic so that it returns if scale values are different.

**Comment by kahmed10** [comment]
Can you add a test case?

**Comment by shivadbhavsar** [comment]
> Can you add a test case?

This pass only really runs with special flags. How would the test cases work for this? 
There are no existing cases for this pass

File: src/targets/gpu/prefuse_ops.cpp (hunk 2 above)
Comment (shivadbhavsar) [reply src/targets/gpu/prefuse_ops.cpp:193]: added

Comment (kahmed10) [reply src/targets/gpu/prefuse_ops.cpp:193]: I was thinking something like this
```suggestion
                if(not std::all_of(....) return;
                scale = s.front();
```

Comment (shivadbhavsar) [reply src/targets/gpu/prefuse_ops.cpp]: But thats inside a visit, that return will not exit the apply method for this matcher

File: test/gpu/prefuse_ops.cpp [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        auto attn =
+            mm->add_instruction(pre_gemm_softmax_gemm{migraphx::make_op("dot"), 2}, x, y, z);
+
+        mm->add_return({attn});
+        return p;
+    };
+
+    migraphx::program p1 = create_program();
+    migraphx::program p2;
+    if(migraphx::gpu::mlir_attention_enabled())
```
Comment (pfultz2) [inline test/gpu/prefuse_ops.cpp:86]: Dont rely on this global setting. Instead add a flag to the `prefuse_ops` class like `enable_attention` that can be set to true during tests. And then you pass that along to the `find_gemm_softmax_gemm` so the matcher will match it.

File: src/targets/gpu/prefuse_ops.cpp (hunk 2 above)
Comment (kahmed10) [reply src/targets/gpu/prefuse_ops.cpp]: oh yeah I saw that too. But looks like you rewrote it and looks fine now.

File: src/targets/gpu/prefuse_ops.cpp [hunk 4]
```diff
@@ -162,36 +162,53 @@ auto is_mlir_gemm()
...
+    return match::make_basic_pred_matcher([=](instruction_ref ins) {
+        if(ins->name() != "dot")
+            return false;
+        return enable_attention;
+    });
+}
+
 struct find_gemm_softmax_gemm
 {
+    bool enable_attention;
```
Comment (pfultz2) [inline src/targets/gpu/prefuse_ops.cpp:176]: Set this to `false`, so its not uninitialized when default constructing.

Comment (shivadbhavsar) [reply src/targets/gpu/prefuse_ops.cpp]: done

**Review by pfultz2 — APPROVED** [review]

**Pull request merged**